### PR TITLE
Execution Client interface implementation

### PR DESCRIPTION
**Pull request by matis755**

### CODE REVIEW

**Review by r4jmond — APPROVED** [review]
guitarr hello, but you could add requirements that the module fulfills, and the test tests

**Review by r4jmond — CHANGES_REQUESTED** [review]
Please replace this implementation with mock and try to find some requirements for this interface.

File: src/execution_client.h [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace ara::exec {
+    class ExecutionClient {
+    public:
+        /**
+        * @brief ReportApplicationState
+        * @details Provides functionality for a Process to report its execution state to Execution Management
+        */
+        virtual void ReportApplicationState(sm::FunctionGroupStateType state) {};
+        virtual ~ExecutionClient() = default;;
```
Comment (r4jmond) [inline src/execution_client.h]: double ;

File: src/state_management.cpp [hunk 2]
```diff
@@ -57,6 +57,10 @@ namespace ara::sm {
         if (stateClient != nullptr) {
             internalState = stateClient->SmGetState();
         }
+        if (executionClient != nullptr) {
+            executionClient->ReportApplicationState(internalState);
+        }
+        triggerOut.SetNotifier(internalState);
```
Comment (r4jmond) [inline src/state_management.cpp]: triggered, this is already called in Work - delete it from there, or from here.

File: src/state_management.cpp [hunk 3]
```diff
@@ -57,6 +57,10 @@ namespace ara::sm {
         if (stateClient != nullptr) {
             internalState = stateClient->SmGetState();
         }
+        if (executionClient != nullptr) {
+            executionClient->ReportApplicationState(internalState);
+        }
```
Comment (r4jmond) [inline src/state_management.cpp:68]: This shouldn't be here - this function is only for updating internal SM state...
but maybe it's good idea to change it - think about it.

File: tests/test_execution_client.cpp [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+                ReportApplicationState(ara::sm::FunctionGroupStateType::On)).Times(testing::AnyNumber());
+
+    mySM.stateClient->SmSetState(ara::sm::FunctionGroupStateType::On);
+
+    // let changes to be applied
+    using namespace std::chrono_literals;
+    std::this_thread::sleep_for(500ms);
+
+    // due to possible GTest stuck while expecting a call
+    mySM.Kill();
```
Comment (r4jmond) [inline tests/test_execution_client.cpp:23]: are u sure that's needed? we are killing SM in teardown function

File: tests/test_execution_client.cpp [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    class MockExecutionClient : public ExecutionClient {
+    public:
+        MOCK_METHOD(void, ReportApplicationState, (sm::FunctionGroupStateType state));
+    };
+}
+
+TEST_F(smTests,  testExecutionClient)
+{
+    ara::exec::MockExecutionClient mockEC;
+    mySM.executionClient = &mockEC;
```
Comment (r4jmond) [inline tests/test_execution_client.cpp]: is this even legal? XD maybe it's better idea to move this to constructor

File: tests/test_execution_client.cpp [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    EXPECT_CALL(mockEC,
+                ReportApplicationState(ara::sm::FunctionGroupStateType::On)).Times(testing::AnyNumber());
+
+    mySM.stateClient->SmSetState(ara::sm::FunctionGroupStateType::On);
```
Comment (r4jmond) [inline tests/test_execution_client.cpp]: You can also add test for report ApplicationState off

**Review by r4jmond — COMMENTED** [review]
small comments

File: src/execution_client.h (hunk 1 above)
Comment (matis755) [reply src/execution_client.h]: Done

File: src/state_management.cpp (hunk 2 above)
Comment (matis755) [reply src/state_management.cpp]: Done

File: src/state_management.cpp (hunk 3 above)
Comment (matis755) [reply src/state_management.cpp:68]: Discussed offline

File: tests/test_execution_client.cpp (hunk 4 above)
Comment (matis755) [reply tests/test_execution_client.cpp:23]: I know, but it is needed here. Without it, EC tests don't work XD

File: tests/test_execution_client.cpp (hunk 5 above)
Comment (matis755) [reply tests/test_execution_client.cpp]: Done

File: tests/test_execution_client.cpp (hunk 6 above)
Comment (matis755) [reply tests/test_execution_client.cpp]: Done

**Review by r4jmond — COMMENTED** [review]
git push

**Pull request merged**